src/CMakeLists.txt

```
@@ -0,0 +1,36 @@
+# libcps
+add_library(libcps cps/env.cpp cps/loader.cpp cps/printer.cpp cps/search.cpp
+                   cps/utils.cpp cps/version.cpp)
+
+# Configure config.hpp
+set(functions unreachable)
+set(headers utility)
+set(BUILTIN_FEATURE_MACROS "")
+foreach(function header IN ZIP_LISTS functions headers)
+  # TODO: remove hard-coded 1 when the use of __builtin_unreachable is refactored out
+  string(TOUPPER ${function} function_upper)
+  set(BUILTIN_FEATURE_MACROS
+      "${BUILTIN_FEATURE_MACROS}\n#define CPS_USE_BUILTIN_${function_upper} 1"
+  )
+endforeach()
+configure_file(cps/config.hpp.in cps/config.hpp)
+
+target_include_directories(libcps PUBLIC .)
+target_include_directories(libcps PUBLIC "${CMAKE_CURRENT_BINARY_DIR}")
+
+# cps-config
+add_executable(cps-config cps-config/main.cpp)
+target_link_libraries(cps-config PRIVATE libcps)
+
+# Dependencies
+find_package(tl-expected 1.0 REQUIRED)
+target_link_libraries(libcps PRIVATE tl::expected)
+
+find_package(fmt 8 REQUIRED)
+target_link_libraries(libcps PRIVATE fmt::fmt)
+
+find_package(jsoncpp 1.9 REQUIRED)
+target_link_libraries(libcps PRIVATE JsonCpp::JsonCpp)
+
+find_package(CLI11 2.1 REQUIRED)
+target_link_libraries(cps-config PRIVATE CLI11::CLI11)
```

```suggestion
find_package(tl-expected REQUIRED)
target_link_libraries(libcps PRIVATE tl::expected)

find_package(fmt REQUIRED)
target_link_libraries(libcps PRIVATE fmt::fmt)

find_package(jsoncpp REQUIRED)
target_link_libraries(libcps PRIVATE JsonCpp::JsonCpp)

find_package(CLI11 REQUIRED)
target_link_libraries(cps-config PRIVATE CLI11::CLI11)
```

Unless we have a specific problem we're trying to work around, it's best to not put specific version information in a build configuration. That info really goes in dependency management logic, for instance in the ubuntu build environment setup logic.

---

I would also move the `add_executable` down to line 34 or so. It's good to keep all the declaration and configuration for a target all together. It's good to have all of the context for how a target is configured all on the same "page" of code, or as close to that as you can get. Just for clarity reasons.

---

Those are the minimum supported versions we use. Meson does have those wired, and I would rather shift left, ie, cmake/meson should fail to configure if you try to use fmt 7 rather than failing at build time, or worse, compiling and then failing in some weird way at runtime.

With my packager hat on, this kind of information helps downstreams too, because ubuntu (for example) can use this information to help decide if they want to take an update, or if the versions they're putting together actually work, or whether they might want to patch cps-config because they want it work with a dependency version they ship we don't support.

---

I only care that the cmake and meson does similar things. I'd be happy to either keep version check in both or neither.

---

Well, the dependency management system (which could be CMake but specifically in FetchContent logic), is more "left" than the build step, and it's the part of the development process responsible for finding versions of dependencies that match, broadly.

Not overspecifying in the build system makes it easy to perform development activities at that layer. One doesn't have to fiddle with CMake and Meson in order to upgrade or downgrade dependencies, especially as a bulk operation.

---

Maybe I'm missing something, but specifying versions as `== 1.2.1` is bad practice, 100% agreed. But saying `>= 1.2` when you in fact rely on features in 1.2 is good practice, since it helps users building your software to get things configured correctly immediately which lowers useless bug reports, and it documents for vendors what you expect. If a software vendor wants to use 1.1 for some reason (lets say Ubuntu in this case), they have mechanisms in place to carry the patch to Meson or CMake that does that as part of their packaging process.

---

If I understand both of you correctly, I think we can agree a minimum version requirement that is not over-specified is beneficial. An over-specified version requirement is not as great. The question at hand then is: are we over-specifying? Does using a lower version of the library actually break anything?

---

In the case of fmt I think it does. I had to do work to make fmt < 9 work (which is why my distro was shipping), and so 
I'd be shocked if 6 or 7 worked as is. I think CLI11 also had issues with 2.0, which is why we have 2.1 there, but I might be misremembering. I know that the library we were using before CLI11 definitely had issues unless you were using specific versions.

---

I believe this remains as the only unresolved thread. @bretbrownjr @dcbaker What do you guys think?

---

Let's leave the version qualifiers in for now so we can move ahead with CMake support. I'm still concerned that they're overspecified inside `CMakeLists.txt`, especially since analogous information will be in the configurations for meson and relevant dependency management systems, but maybe we can follow up on that later.